Source/CarthageKit/Git.swift

```diff
@@ -294,83 +294,52 @@ public func checkoutRepositoryToDirectory(_ repositoryFileURL: URL, _ workingDir
 /// repository, but without any Git metadata.
 public func cloneSubmoduleInWorkingDirectory(_ submodule: Submodule, _ workingDirectoryURL: URL) -> SignalProducer<(), CarthageError> {
 	let submoduleDirectoryURL = workingDirectoryURL.appendingPathComponent(submodule.path, isDirectory: true)
+
+	func repositoryCheck<T>(attempt closure: () throws -> T, reasonForFailure: String) -> Result<T, CarthageError> {
+		do {
+			return .success(try closure())
+		} catch let error as NSError {
+			return .failure(
+				.repositoryCheckoutFailed(workingDirectoryURL: submoduleDirectoryURL, reason: reasonForFailure, underlyingError: error)
+			)
+		}
+	}
+
+	// swiftlint:disable switch_case_on_newline
 	let purgeGitDirectories = FileManager.default.reactive
 		.enumerator(at: submoduleDirectoryURL, includingPropertiesForKeys: [ .isDirectoryKey, .nameKey ], catchErrors: true)
-		.flatMap(.merge) { enumerator, url -> SignalProducer<(), CarthageError> in
-			var name: String?
-			do {
-				name = try url.resourceValues(forKeys: [ .nameKey ]).name
-			} catch let error as NSError {
-				return SignalProducer(
-					error: CarthageError.repositoryCheckoutFailed(
-						workingDirectoryURL: submoduleDirectoryURL,
-						reason: "could not enumerate name of descendant at \(url.path)",
-						underlyingError: error
-					)
-				)
+		.attemptMap { enumerator, url -> Result<(), CarthageError> in
+			switch repositoryCheck(attempt: {
+				try url.resourceValues(forKeys: [ .nameKey ]).name
+			}, reasonForFailure: "could not enumerate name of descendant at \(url.path)") {
+			case .failure(let error): return .failure(error)
+			case .success(let name) where name != ".git": return .success(())
+			default: break // proceed to the below `repositoryCheck`
 			}
 
-			if name != ".git" {
-				return .empty
-			}
-
-			var isDirectory: Bool?
-			do {
-				isDirectory = try url.resourceValues(forKeys: [ .isDirectoryKey ]).isDirectory
-				if isDirectory == nil {
-					return SignalProducer(
-						error: CarthageError.repositoryCheckoutFailed(
-							workingDirectoryURL: submoduleDirectoryURL,
-							reason: "could not determine whether \(url.path) is a directory",
-							underlyingError: nil
-						)
-					)
+			return repositoryCheck(attempt: {
```

Comment:
If we put the string first, we could use trailing closure syntax here, which I think would be more readable.

```swift
// the error could add the "could not " before the reason string here
return repositoryCheck("determine whether \(url.path) is a directory") {
        switch try url.resourceValues(forKeys: [ .isDirectoryKey ]).isDirectory {
        case .none:
            throw NSError() // this bogus error will probably never be seen
        case .some(let bool):
            return bool
        }
     }
     .flatMap { (isDirectory: Bool) in
         if isDirectory { enumerator.skipDescendants() }
             return repositoryCheck("remove \(url.path)") {
                 try FileManager.default.removeItem(at: url)
             }
         }
     }
```